main/src/ca/uwaterloo/flix/language/phase/unification/BoolMinimization.scala

```diff
@@ -0,0 +1,482 @@
+/*
+ *  Copyright 2022 Jonathan Lindegaard Starup
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.unification
+
+import ca.uwaterloo.flix.language.ast.{Kind, Rigidity, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+import scala.annotation.tailrec
+
+object BoolMinimization {
+
+  sealed trait Formula
+
+  object Formula {
+
+    case object True extends Formula
+
+    case object False extends Formula
+
+    case class Var(v: Type.KindedVar) extends Formula
+
+    case class Not(term: Formula) extends Formula
+
+    case class And(terms: List[Formula]) extends Formula
+
+    case class Or(terms: List[Formula]) extends Formula
+
+  }
+
+  /**
+    * Sort terms `True` -> `False` -> `Var`s (by name then negation) ->
+    * complex negation -> conjunctions -> disjunctions.
+    */
+  private def sortTerms(terms: List[Formula]): List[Formula] = {
+    import Formula._
+    terms.sortWith {
+      case (True, True) => false
+      case (True, False | Var(_) | Not(_) | And(_) | Or(_)) => true
+
+      case (False, True | False) => false
+      case (False, Var(_) | Not(_) | And(_) | Or(_)) => true
+
+      case (Var(_), True | False) => false
+
+      case (Var(v1), Var(v2)) => v1 < v2
+      case (Var(v1), Not(Var(v2))) if v1 == v2 => true
+      case (Var(v1), Not(Var(v2))) => v1 < v2
+      case (Not(Var(v1)), Var(v2)) if v1 == v2 => false
+      case (Not(Var(v1)), Var(v2)) => v1 < v2
+      case (Not(Var(v1)), Not(Var(v2))) => v1 < v2
+      case (Var(_), Not(_) | And(_) | Or(_)) => true
+
+      case (Not(_), True | False | Var(_)) => false
+      case (Not(_), Not(_)) => false
+      case (Not(_), And(_) | Or(_)) => true
+
+      case (And(_), True | False | Var(_) | Not(_)) => false
+      case (And(_), And(_)) => false
+      case (And(_), Or(_)) => true
+
+      case (Or(_), True | False | Var(_) | Not(_) | And(_) | Or(_)) => false
+    }
+  }
+
+  /**
+    * Reduces things like `y ∨ y` to `y` or `¬x ∧ ¬x` to `x`. Returns `None` if
```

Comment:
I think we want one function (which internally uses three mutable sets) to compute:

- set of vars
- set of negated vars
- list of the rest.